### PR TITLE
Rewrite `array @> array` and `array <@ array` in sql_expr_to_logical_expr

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Part of #8506 #10534 

## Rationale for this change

Move the rewrite from optimizer to sql crate

I keep the function rewrite trait even it is not used anywhere in datafusion now.

### CODE REVIEW

File: datafusion/sql/tests/sql_integration.rs [hunk 1]
```diff
@@ -1225,22 +1225,6 @@ fn select_binary_expr_nested() {
     quick_test(sql, expected);
 }
 
-#[test]
-fn select_at_arrow_operator() {
```
Comment (alamb) [inline datafusion/sql/tests/sql_integration.rs:1229]: Should we add an explain test?

**Review by alamb — APPROVED** [review]
Thanks @jayzhan211  -- this looks good to me

Can you remind me why this method is preferrable to the FunctionRewrite idea we previously had? 

I feel this PR and changes are related to the one from @samuelcolvin  on   https://github.com/apache/datafusion/pull/11137 -- where the idea is to support SQL syntax with JSON operators without putting the implementation into SQL

**Comment by alamb** [comment]
Here is an alternate idea https://github.com/apache/datafusion/pull/11168

File: datafusion/sql/tests/sql_integration.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/sql/tests/sql_integration.rs:1229]: I don't think it is needed since we don't really care about how it looks in plan.

**Comment by jayzhan211** [comment]
> Thanks @jayzhan211 -- this looks good to me
> 
> Can you remind me why this method is preferrable to the FunctionRewrite idea we previously had?
> 
> I feel this PR and changes are related to the one from @samuelcolvin on #11137 -- where the idea is to support SQL syntax with JSON operators without putting the implementation into SQL

From your comment here :) https://github.com/apache/datafusion/issues/10534#issuecomment-2113211517

File: datafusion/sql/tests/sql_integration.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/sql/tests/sql_integration.rs:1229]: Maybe we can make sure `array_has_all` is applied 🤔

File: datafusion/physical-expr-common/src/datum.rs [hunk 2]
```diff
@@ -145,36 +145,3 @@ fn compare_op_for_nested(
         Ok(BooleanArray::new(values, nulls))
     }
 }
-
-#[cfg(test)]
-mod tests {
```
Comment (jayzhan211) [inline datafusion/physical-expr-common/src/datum.rs:150]: forgot to remove in previous PR

**Pull request merged**